From d37da3f924763324759abf9ce8c5159d1fe71c5b Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Wed, 28 Feb 2018 14:44:25 +0800
Subject: [PATCH] drm/rockchip: vop: close all win when enable vop

Because vop is unstable when update output timing,
so we close all win make sure stable.

Change-Id: Ifd4e60a41c667426d40e0d57e5180034bab9d5e7
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop.c | 18 +++++++-----------
 1 file changed, 7 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
index ea4678d31bcf..7e8640bbf605 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
@@ -2439,6 +2439,13 @@ static void vop_crtc_enable(struct drm_crtc *crtc)
 	mutex_lock(&vop->vop_lock);
 	vop_initial(crtc);
 
+	VOP_CTRL_SET(vop, standby, 0);
+	if (vop_crtc_mode_update(crtc)) {
+		vop_disable_all_planes(vop);
+		vop->mode_update = true;
+		DRM_DEV_INFO(vop->dev, "Update mode to %d*%d, close all win\n",
+			     hdisplay, vdisplay);
+	}
 	VOP_CTRL_SET(vop, dclk_pol, 1);
 	val = (adjusted_mode->flags & DRM_MODE_FLAG_NHSYNC) ?
 		   0 : BIT(HSYNC_POSITIVE);
@@ -2553,20 +2560,9 @@ static void vop_crtc_enable(struct drm_crtc *crtc)
 	VOP_CTRL_SET(vop, cabc_global_dn_limit_en, 1);
 	VOP_CTRL_SET(vop, win_csc_mode_sel, 1);
 
-	if (vop_crtc_mode_update(crtc)) {
-		vop_disable_all_planes(vop);
-		vop->mode_update = true;
-		DRM_DEV_INFO(vop->dev, "Update mode to %d*%d, close all win\n",
-			      hdisplay, vdisplay);
-	}
-
 	clk_set_rate(vop->dclk, adjusted_mode->crtc_clock * 1000);
 
 	vop_cfg_done(vop);
-	/*
-	 * enable vop, all the register would take effect when vop exit standby
-	 */
-	VOP_CTRL_SET(vop, standby, 0);
 
 	enable_irq(vop->irq);
 	drm_crtc_vblank_on(crtc);
-- 
2.35.3

